### DISCUSSION TITLE
Dom 셀렉팅에 대하여...

### DISCUSSION THREAD

**Discussion by lokba** [woowacourse-study/2022-code-review-study#13] [Lotto] [answered]
크루분들의 코드에서, 아래와 같은 코드를 꽤나 보았습니다.

```javascript
//util
export const selectDom = (selector, parent = document) => parent.querySelector(selector);
```

위 유틸함수의 목적은 제가 이해한 바로는 다음과 같습니다.

```javascript
//1번
const childElement = document.getElementById("child");

//2번
const childElement = parentElement.getElementById("child");
```

제 질문은 다음과 같습니다.

Q1. 1번과 2번의 dom을 탐색하는데 뭐가 더 좋은 방식일까요?

Q2. 만약 2번이라고 생각하신다면, util을 저런식으로 만드는 것에 대해서는 어떻게 생각하시나요?

Q3. 결국에 저 유틸의 목적이 dom 접근 최적화인데요. 한 리뷰어는 다음과 같은 코멘트를 달았습니다. 아래에 대해서는 어떻게 생각하시나요?
> 저는 최적화에 대해서 측정이 가능하거나 시간복잡도로 계산이 될때 하는게 의미있다고 생각하는편입니다. 참고링크 [여기](https://stackify.com/premature-optimization-evil/)

**Comment by moonheekim0118** [comment]
1. 저는 2번이 더 좋은 방식이라고 생각합니다. 물론 속도상의 이점도 있겠죠? 그런데 3번 리뷰어님의 말씀처럼 저희가 판단하기는 어려울 것 같아요. 그런데 저는 성능보다는 한 컴포넌트 내부에서 Element들의 부모-자식 관계를 더 명시적으로 나타내주어서 직관적이라고 생각했어요. 예를들면 FormView 라는 컴포넌트 내부에서  ` this.Input = selectDom('#input', this.container)` 이렇게 찾아준다면, 명확하게 '아 ! 이 Input은 현재 컴포넌트(FormView)에 속해있는 자식이구나!'를 알 수 있으니까요? 뭔가 개발자로서 실수(다른 Form에 속해있는 Input을 셀렉팅한다던지..)가 적어지지 않을까 생각했습니다.
2. 좋은 방법이라고 생각해요. 록바는 혹시 해당 util에서 잘못된 부분이 있다고 생각하는 점이 있을까요? 저는 저런식으로 util을 작성해왔어서 다른 의견이 궁금하네요 ~~
3. 이 답변은 1번에 속해있다고 생각합니다. ㅎㅎ저는 접근최적화보다는 직관성을 높여주는 목적이 더 크다고 생각하는 편이예요.

**Reply by lokba** [reply]
좋은 답변 감사합니다. 호프

> 속도상의 이점뿐만 아니라, 명시적이기에 더 직관적이다.

오 그렇게 볼 수도 있겠네요. 저는 해당 util 코드를 처음 접하였기에 속도상의 이점만 고려하여 생각을 하였는데, 직관적으로도 생각할 수 있군요.

<br>

> 록바는 혹시 해당 util에서 잘못된 부분이 있다고 생각하는 점이 있을까요?

제 질문이 살짝 모호했네요. util에 있는 함수 자체는 문제가 없다고 생각이 드는데, 저렇게 함수를 만들어서 쓰는게 어떻게 생각하는지 궁금해서 물어봤습니다.

**Reply by moonheekim0118** [reply]
저야말로 질문을 잘못 이해했었네요~ 저는 개인적으로 저렇게 util 만드는걸 선호합니다. 매번 쿼리셀렉터 불러오는 것 보다 괜찮다고 생각하는 편이예요. 근데 자칫잘못하면 기존 메서드(querySelector)의 스펙을 변경하여 사용 할 수 있어서 혼선을 줄 수도 있겠네요..

**Comment by wonsss** [comment, marked as answer]
레이싱게임 리뷰중에서 관련 글을 본 적 있습니다.

selector 탐색을 root에서부터 시작하는 경우가 많은데, 이럴 경우 탐색하면서 매칭되는 첫 번째 element를 반환하게 됩니다. 이런 방식으로는 프로젝트 덩치가 커질수록 다음과 같은 두 가지 이슈가 예상된다고 하는데요.

(1) 모든 element를 root에서부터 탐색하기 때문에 element 깊이가 깊어질수록 탐색 속도 저하
(2) 유일한 className이 아니면 의도하지 않은 element 반환
이러한 이슈를 방지하기 위해, document root부터 탐색이 아니라 특정 element 부터 탐색하는 방법을 고민해보기 위해 2번과 같은 코드가 나온 것 같습니다.
